### PR TITLE
observe everything

### PR DESCRIPTION
**Pull request by fwkoch**
- `everything` sentinel now used to specify when all properties should be observed.

Minor changes:
- `reset` was made private `_reset`
- listeners/validators are now stored in `OrderedDicts` to give some consistency to the order they are called in (order is still ambiguous within a class, but inherited order is kept consistent, ie `HasProperties` validators are called first).

### CODE REVIEW

File: properties/base.py [hunk 1]
```diff
@@ -42,27 +43,25 @@ def __new__(mcs, name, bases, classdict):                                  #pyli
...
-                })
-            if hasattr(base, '_class_validators'):
-                _class_validators.update({
-                    k: v for k, v in iteritems(base._class_validators)
-                    # drop ones which are no longer validators
-                    if not (k not in validator_dict and k in classdict)
-                })
+            if not (hasattr(base, '_props') and
+                    hasattr(base, '_prop_observers') and
+                    hasattr(base, '_class_validators')):
```
Comment (bsmithyman) [inline properties/base.py]: If you want you can use `all` here

File: properties/base.py [hunk 2]
```diff
@@ -85,6 +84,8 @@ def __new__(mcs, name, bases, classdict):                                  #pyli
 
         # Ensure observed names are valid
         for key, hand in iteritems(observer_dict):
+            if hand.names is utils.everything:
+                continue
```
Comment (bsmithyman) [inline properties/base.py]: Cool

File: properties/base.py [hunk 3]
```diff
@@ -191,11 +192,10 @@ def __init__(self, **kwargs):
                     continue
                 if callable(val):
                     val = val()
-                self._props[key].validate(self, val)
-                self._backend[key] = val
+                self._backend[key] = self._props[key].validate(self, val)
```
Comment (bsmithyman) [inline properties/base.py]: Right, so it is coerced?

File: properties/handlers.py [hunk 4]
```diff
@@ -6,10 +6,16 @@
...
 from six import string_types
 
+from .utils import everything
+
 
 def _set_listener(instance, obs):
     """Add listeners to a Properties class instance"""
-    for name in obs.names:
+    if obs.names is everything:
+        names = list(instance._props)
```
Comment (bsmithyman) [inline properties/handlers.py]: This is wonderfully Pythonic, but I wonder if `list(instance._props.keys())` might be less magic & more clear.

File: properties/base.py (hunk 3 above)
Comment (fwkoch) [inline properties/base.py]: Yes, I think validating defaults but not coercing was an oversight.

File: properties/base.py (hunk 1 above)
Comment (fwkoch) [inline properties/base.py]: Good idea - that would make it more readable.

File: properties/handlers.py (hunk 4 above)
Comment (fwkoch) [inline properties/handlers.py]: Hmm, I think I might disagree here. I don't think this behaviour is too magic; it is the python-future recommended way to make a list of keys (http://python-future.org/compatible_idioms.html#dict-keys-values-items-as-a-list). Additionally, it is assigned to `names` - maybe that will cue someone who is unclear that this is a list of names?

**Pull request merged**